[Common-Algorithms] Implemented and tested common algorithms in Chapel. Also implemented some parallel algorithms

rohanbadlani opened the pull request:
Hi,

During the learning phase of Chapel, I have implemented a few commonly used algorithms. In these algorithms I have tried to cover as many constructs as possible. However, a few features like synchronization could not be exemplified through these implementations. I have even implemented a few parallel algorithms like parallel mergesort, quicksort, etc, keeping in mind that even if the user of the Chapel language is thinking from a sequential standpoint, he/she can easily improve performance of his tasks by using constructs like begin, cobegin, forall, etc. Thus, these examples illustrate the uniqueness of Chapel. All of these have been tested using the Chapel Testing System.

I hope you find these examples good enough to include as samples in the project.

While implementing these and many other algorithms (especially graph based algorithms), I have encountered a few bugs in Chapel and a few features that I as a programmer using Chapel would have preferred. I would be reporting them through the bug-reporting platform.

Regards,
Rohan.

bradcray commented:
[To the core Chapel developer who reviews this PR -- think about whether some of these tests might be good candidates for test/release/examples/programs/ or /patterns/ ]

Hi @rohanbadlani : Since you're already comfortable with the Chapel testing system, you could file your bugs directly as "future" tests, which is how we tend to track such issues on the automated regression testing side of the project.  For more details, see:
 https://github.com/chapel-lang/chapel/blob/master/doc/developer/bestPractices/TestSystem.rst

and more specifically:

https://github.com/chapel-lang/chapel/blob/master/doc/developer/bestPractices/TestSystem.rst#futures-a-mechanism-for-tracking-bugs-feature-requests-etc

rohanbadlani commented:
Hi,

I have added the bugs I faced while developing sequential programs in Chapel. I have tested and added all the bugs as 'future' tests with corresponding bug file (chpl program) alongwith .future, .good and .bad files. 

I have also added bug-description file explaining the bugs. This also included my system specifications as asked in the Bug Requests Page. 

While working on parallel constructs such as begin, forall, cobegin and coforall: I was facing a few bugs. I'll be adding them as 'future' tests in a day or two.

I am new to the Chapel Community. Hence, if any of the added bugs / feature requests are known beforehand or already resolved, kindly pardon me.

Regards,
Rohan.

benharsh commented on test/bugs-reported/bug1.chpl:
```diff
@@ -0,0 +1,100 @@
+
```
Could we shorten this program down to something simpler? I think this could be done in a couple of lines:

```chpl
var x = 5;
writeln(x++);
```

benharsh replied:
On second thought I think this may be more appropriate as a JIRA issue, as there will probably be some discussion. Here's a link to our JIRA: https://chapel.atlassian.net/projects/CHAPEL/issues/

rohanbadlani replied:
Added this as a JIRA issue available here: https://chapel.atlassian.net/browse/CHAPEL-196
Added this long program to show the motivation as to why the programmer may need this. 
Will remove this bug as a "future" test as the JIRA issue has been created.

benharsh commented on test/bugs-reported/bug2.chpl:
```diff
@@ -0,0 +1,102 @@
+
```
Same here: could be much shorter and easier to understand.

rohanbadlani replied:
I agree, the code is irrelevant to the issue being addressed here. Updated bug2.chpl, bug2.good, bug2.bad accordingly. Made the code more shorter and relevant.

benharsh commented on test/bugs-reported/bug3.chpl:
```diff
@@ -0,0 +1,8 @@
+proc max(a:int, b:int): int{
```
I believe this would also be more appropriate as a JIRA issue.

In case you didn't know, you can accomplish behavior with if/then:
```chpl
proc max(a:int, b:int) {
  return if a > b then a else b;
}
```

rohanbadlani replied:
Created a JIRA Issue for this: https://chapel.atlassian.net/browse/CHAPEL-197

Will remove this as a "future" test.

rohanbadlani replied:
Yeah, I noticed the single statement if..else. But, I felt, in some scenarios as explained in the JIRA issue, the conditional operators may be more beneficial. Pardon me, if you still feel that if..else can handle all these.

benharsh commented on test/bugs-reported/bug4.future:
```diff
@@ -0,0 +1,28 @@
+bug: The List Module does not support accessing elements of a list based on index. I feel that this should be supported.
```
Yikes! This one's embarassing...

I'd advocate for splitting this up into two smaller programs/futures.

I don't think we want to allow indexing through the ".these" iterator, but we definitely shouldn't emit an internal compiler error.

rohanbadlani replied:
I have split bug4 into 2 bugs: bug8.chpl and bug9.chpl with respective future, good and bad files.

I agree that .these() iterator should not be used for accessing list elements by index. But, it should throw an error like "Iterator does not allow indexed access."

benharsh replied:
Thanks for splitting this into two much simpler files. Can the bug4 files be removed?

rohanbadlani replied:
Removed bug4.

benharsh commented on test/bugs-reported/bug5.chpl:
```diff
@@ -0,0 +1,71 @@
+use List;
```
I think this would be better as a JIRA issue.

rohanbadlani replied:
Agreed. A new JIRA Issue is created here: https://chapel.atlassian.net/browse/CHAPEL-198
Will be removing this as a future test.

benharsh commented on test/bugs-reported/bug6.chpl:
```diff
@@ -0,0 +1,14 @@
+use Queue;
```
I believe this is an error because the compiler fails to find a file named "Queue.chpl". @lydia-duncan , can you chime in here? Should this be a feature request in JIRA instead?

lydia-duncan replied:
I agree that this is because the compiler hasn't found a file named "Queue.chpl".  The solution is to name the file with Queue in it on the compilation line (i.e. add a .compopts file with the additional module file, plus probably --main-module bug6 or something).  I believe this behavior is okay, since there could be multiple files in the same directory which define a module by that name and the compiler wouldn't know which one to choose if it had to look into all files at this directory scope.

benharsh replied:
Thanks lydia. For now I think it makes sense to remove this bug.

rohanbadlani replied:
I did not know that multiple files can define same module and the programmer can choose which to use. Thanks @lydia-duncan for clarifying this. However, I am still unable to compiler this. Can you please direct me to a resource that explains how to do this?

@benharsh Removing this bug.

lydia-duncan replied:
So far as I can tell, the reason this program would still fail to compile is because of the bug in the file with the Queue module in it.  Having a clean Queue module to work with would resolve the issue.

I poked around to find a good summary description of using outside modules, but the ones I found all seem to be too high level.  I have an action item to write a program for the release/examples/primers directory covering use statements, which I intend to work on shortly.

rohanbadlani replied:
Ohh..I see. Cool. I'll definitely refer to the release examples for use statement later.

benharsh commented on test/bugs-reported/bug7.chpl:
```diff
@@ -0,0 +1,15 @@
+use Graph;
```
I think this would be more appropriate as a JIRA issue that included a description of the kinds of graph representations you would like to see.

rohanbadlani replied:
Created a JIRA issue available here: https://chapel.atlassian.net/projects/CHAPEL/issues/CHAPEL-199

Removing this as a "future" test.

benharsh commented:
Despite the number of suggested changes, we do appreciate your interest and willingness to bring up these issues with us :smile: 

@ben-albrecht will continue reviewing some of the other non-bug additions.

ben-albrecht commented on test/common-algorithms/dynamic-programming/sequential/coin_change.chpl:
```diff
@@ -0,0 +1,31 @@
+//Time Complexity: O(mn)
+proc count(S, m:int, n:int):int{
+	// table[i] will be storing the number of solutions for
+	// value i. We need n+1 rows as the table is consturcted
```
s/consturcted/constructed/

rohanbadlani replied:
Sorry for the typo.

ben-albrecht commented on test/common-algorithms/dynamic-programming/sequential/coin_change.chpl:
```diff
@@ -0,0 +1,31 @@
+	var table: [1..n+1] int = 0;
+	
+	// Initialize all table values as 0
+	table = 0;
```
line 6 already accomplishes this.

rohanbadlani replied:
Yeah...removing line 9.

ben-albrecht commented on test/common-algorithms/dynamic-programming/sequential/coin_change.chpl:
```diff
@@ -0,0 +1,31 @@
+	// Pick all coins one by one and update the table[] values
+	// after the index greater than or equal to the value of the
+	// picked coin
+	var i:int;
```
In Chapel, a for loop statement declares new variables for the scope of the loop, so this is not necessary.

ben-albrecht commented on test/common-algorithms/dynamic-programming/sequential/coin_change.chpl:
```diff
@@ -0,0 +1,31 @@
+	// picked coin
+	var i:int;
+	for i in (1..m){
+		var j:int;
```
Same with `j`

ben-albrecht commented on test/common-algorithms/dynamic-programming/sequential/coin_change.chpl:
```diff
@@ -0,0 +1,31 @@
+	// after the index greater than or equal to the value of the
+	// picked coin
+	var i:int;
+	for i in (1..m){
```
no parentheses necessary here.

ben-albrecht commented on test/common-algorithms/dynamic-programming/sequential/coin_change.chpl:
```diff
@@ -0,0 +1,31 @@
+		for j in (S(i)..n){
+			table(j) += table(j-S(i)+1);
+ 		}
+	}
```
I would rewrite this loop as:

```chapel
for (i, j) in zip(1..m, S[i]..n) {
  table[j] += table[j - S[i] + 1];
}
// Note: I support [brackets] over (parens) for array access stylistically, but it's ultimately up to you.
```

rohanbadlani replied:
Since the inner loop range depends on the outer loop, I think zip wouldn't work. I tried it and as expected, the following error occurs.
error: zippered iterations have non-equal lengths

Also, for the loop suggested above, when I try to compiler, it requires i and j to be declared. I am thinking that maybe this is a bug. What do you think?

Nevertheless, I have written a nested loop and removed looping variable declarations.

ben-albrecht commented on test/common-algorithms/dynamic-programming/sequential/coin_change.chpl:
```diff
@@ -0,0 +1,31 @@
+var arr: [1..3] int = (1, 2, 3);
+var m : int = 3;
+
+writeln("The count = ", count(arr, m, 4));
```
Just curious, why declare `m`, but not `n`?

ben-albrecht commented on test/common-algorithms/dynamic-programming/sequential/edit_distance.chpl:
```diff
@@ -0,0 +1,42 @@
+proc min(x:int, y:int):int{
+	if(x > y){
```
Just so you know, the short-hand alternative exists:

```chapel
if x > y then return y;
else return x;
```

rohanbadlani replied:
Yeah...I agree, my bad. :+1:

ben-albrecht commented on test/common-algorithms/dynamic-programming/sequential/fibonacci.chpl:
```diff
@@ -0,0 +1,25 @@
+proc fib(n:int):int{
```
There are actually several implementations of Fibonacci already in the test directory (grep for `fib`)

rohanbadlani replied:
Sorry, didn't know this. I should have done a thorough search. Removing this.

ben-albrecht commented on test/common-algorithms/dynamic-programming/sequential/knapsack_recursive.chpl:
```diff
@@ -0,0 +1,36 @@
+proc max(a:int, b:int): int{
```
Use [built-in max](http://chapel.cray.com/docs/master/modules/internal/UtilMisc_forDocs.html#UtilMisc_forDocs.max)

rohanbadlani replied:
Updated.

ben-albrecht commented on test/common-algorithms/dynamic-programming/sequential/edit_distance.chpl:
```diff
@@ -0,0 +1,42 @@
+	}
+}
+
+proc min(x:int,y:int,z:int):int{
```
Use [built-in min](http://chapel.cray.com/docs/master/modules/internal/UtilMisc_forDocs.html#UtilMisc_forDocs.min)

rohanbadlani replied:
Updated.

ben-albrecht commented on test/common-algorithms/dynamic-programming/sequential/longest_common_subsequence.chpl:
```diff
@@ -0,0 +1,25 @@
+proc max(a:int, b:int): int{
```
use built-in version

rohanbadlani replied:
Updated.

ben-albrecht commented on test/common-algorithms/selection/sequential/sequential_quickselect.chpl:
```diff
@@ -0,0 +1,65 @@
+//var arr: [1..10] int = (8, 14, -8, -9, 5, -9, -3, 0, 17, 19);
```
left over comment?

rohanbadlani replied:
Yikes. Sorry, removing unnecessary debug comments.

ben-albrecht commented on test/common-algorithms/sorting/parallel/parallel_quicksort.chpl:
```diff
@@ -0,0 +1,62 @@
+			quicksort(arr, pivot_index+1, right);
+		}
+		
+		/*
```
left over comments?

ben-albrecht commented on test/common-algorithms/sorting/sequential/sequential_mergesort.chpl:
```diff
@@ -0,0 +1,100 @@
+var arr: [1..100] int = (8, 14, -8, -9, 5, -9, -3, 0, 17, 19, 8, 14, -8, -9, 5, -9, -3, 0, 17, 19, 8, 14, -8, -9, 5, -9, -3, 0, 17, 19, 8, 14, -8, -9, 5, -9, -3, 0, 17, 19, 8, 14, -8, -9, 5, -9, -3, 0, 17, 19, 8, 14, -8, -9, 5, -9, -3, 0, 17, 19, 8, 14, -8, -9, 5, -9, -3, 0, 17, 19, 8, 14, -8, -9, 5, -9, -3, 0, 17, 19, 8, 14, -8, -9, 5, -9, -3, 0, 17, 19, 8, 14, -8, -9, 5, -9, -3, 0, 17, 19);
+
+writeln("Input Array:");
+for i in (1..100){
```
The following can accomplish the same thing:

```chapel
writeln(arr);
```

rohanbadlani replied:
Updated. Also used the built-in swap operator in respective places.

ben-albrecht commented on test/common-algorithms/sorting/sequential/sequential_quicksort.chpl:
```diff
@@ -0,0 +1,60 @@
+		quicksort(arr, left, pivot_index-1);
+		quicksort(arr, pivot_index+1, right);
+		
+		
```
left-over comments

rohanbadlani replied:
Removed unnecessary comments.

ben-albrecht commented:
Hi @rohanbadlani  - Thanks for the contributions.

For the contributed algorithms, there are some existing implementations in the test hierarchy. A new implementation would need to be significantly different in order to justify adding it to the test system. I would remove the following implementations, because they already exist:

* `common-algorithms/dynamic-programming/sequential/fibonacci.chpl`
    * Already exists: [release/examples/primers/learnChapelInYMinutes.chpl](https://github.com/chapel-lang/chapel/blob/master/test/release/examples/primers/learnChapelInYMinutes.chpl#L404)
* `common-algorithms/dynamic-programming/sequential/longest_common_subsequence.chpl`
    * Already exists: [trivial/diten/lcs/lcs.chpl](https://github.com/chapel-lang/chapel/blob/master/test/trivial/diten/lcs/lcs.chpl)

The sort implementations already exist in the [Sort package module](http://chapel.cray.com/docs/master/modules/packages/Sort.html). However, we might be able to accept them as pedagogical implementations. I'd be curious to hear what others think - in particular: @bradcray , @benharsh , @lydia-duncan .

* `common-algorithms/sorting/parallel/parallel_mergesort.chpl`
* `common-algorithms/sorting/parallel/parallel_quicksort.chpl`
* `common-algorithms/sorting/sequential/sequential_mergesort.chpl`
* `common-algorithms/sorting/sequential/sequential_quicksort.chpl`

In future pull requests, I recommend breaking up contributions into smaller chunks. Up to ~200 lines of code is ideal for reviewing, but it really depends on the size of the feature. For this PR, I would even go as far as recommending submitting each bug/feature request and new algorithm as an individual PR to ensure a quicker and higher quality review over each one. I think this PR as it stands will be fine, but we may ask you to break up future PRs.

lydia-duncan commented on test/common-algorithms/sorting/parallel/parallel_mergesort.chpl:
```diff
@@ -0,0 +1,102 @@
+//Second subarray is arr[m+1..r]
+
+proc merge(arr, left:int, mid:int, right:int){
+	var i: int = 0, j: int = 0, k: int = 0;
```
Note that there's a short hand for these declarations:
`var i, j, k: int = 0;` See section 8.1.4 (Multiple Variable Declarations) of the Chapel [spec](http://chapel.cray.com/spec/spec-0.98.pdf).
And that additionally, since 0 is an integer literal value, you don't need to give the type specifier in the first place (though you're certainly welcome to do so!)

bradcray replied:
This is also one of the few parts of the language to be covered in the user's guide so far!  :)

http://chapel.cray.com/docs/master/users-guide/base/variableDeclarations.html

rohanbadlani replied:
Sorry, missed the shorthand. Its great to multiple variable declarations with initialization support. :)

lydia-duncan commented on test/common-algorithms/sorting/parallel/parallel_mergesort.chpl:
```diff
@@ -0,0 +1,102 @@
+}
+
+//parallel mergesort algorithm
+//can improve performance by adding heuristics -> When the array size becomes small such that creaing a task takes more time than compute time for sorting it, then dont spawn task, rather do it sequentially.
```
Typo on "creating".  Also, might be nice to wrap these lines to 80 characters, it goes off the screen here.

lydia-duncan commented on test/common-algorithms/sorting/parallel/parallel_quicksort.chpl:
```diff
@@ -0,0 +1,62 @@
+            		i+=1;
+            		var temp : int = arr(j);
+			arr(j) = arr(i);
+			arr(i) = temp;
```
We actually have a swap operator, `<=>` (listed under [Assignments](http://chapel.cray.com/spec/quickReference.pdf)), that you can use instead of the explicit swap.

lydia-duncan commented on test/common-algorithms/sorting/parallel/parallel_quicksort.chpl:
```diff
@@ -0,0 +1,62 @@
+    	}
+    	var temp : int = arr(i+1);
+	arr(i+1) = arr(right);
+	arr(right) = temp;
```
Same note here

lydia-duncan commented on test/common-algorithms/sorting/sequential/sequential_mergesort.chpl:
```diff
@@ -0,0 +1,100 @@
+//First subarray is arr[l..m]
+//Second subarray is arr[m+1..r]
+
+proc merge(arr, left:int, mid:int, right:int){
```
This is identical code to the parallel case, correct?  If so, you could merge the two and add a flag/threshold for when to use parallel or sequential (the threshold could also be used to show the heuristic you were describing, killing two birds with one stone)

lydia-duncan commented on test/common-algorithms/sorting/sequential/sequential_mergesort.chpl:
```diff
@@ -0,0 +1,100 @@
+//Second subarray is arr[m+1..r]
+
+proc merge(arr, left:int, mid:int, right:int){
+	var i: int = 0, j: int = 0, k: int = 0;
```
Be sure to make the same changes to this file as you do for the parallel one, and similarly for the quicksort pair

rohanbadlani commented:
Hi,

Thanks for all the suggestions and reviewing the code. All suggestions were valid and the code has been modified accordingly wherever applicable. I request you to kindly do a final code review on this.

@ben-albrecht I completely agree with you that I should have created multiple Pull requests for this. From the next time around, I'll definitely break the PRs based on features implemented. 

Thanks for the suggestions and having the patience to review it all together.

Regards,
Rohan.

benharsh commented on test/bugs-reported/bug-descriptions:
```diff
@@ -0,0 +1,143 @@
+*****************************************
```
I don't think this file is necessary given the contents of the .future files.

Also, I think we would prefer a different directory for these bugs. While you might be able to find an appropriate existing directory, it might be easiest to create test/users/rohanbadlani/bugs/ instead.

rohanbadlani replied:
Removed bug-descriptions. Placed all bugs in test/users/rohanbadlani/bugs. Removed bug-descriptions/

benharsh commented on test/bugs-reported/bug9.chpl:
```diff
@@ -0,0 +1,16 @@
+writeln("");
+
+writeln("Accessing list by index using the .these() iterator.");
+//writeln(mylist.these(2));
```
I think this line should be uncommented, right?

rohanbadlani replied:
Yikes. Sorry, removed the comment.

ben-albrecht commented:
Hi @rohanbadlani - I think this is nearly ready to merge. 

I think a good home for the contents of `test/common-algorithms/*/*/` would be a new directory: `test/patterns/`.

This can serve as a staging area for programs that we'd like to some day ship with the release in `test/release/examples/patterns`, once they have received enough attention, e.g. parallel versions.

Once this change is in, I think we will be ready to merge. Thanks for the contributions!

rohanbadlani commented:
Hi @ben-albrecht - Cleaned the repo and moved contents of common-algorithms/ to patterns/

Thanks for thoroughly reviewing the code.

buddha314 commented:
I see you did a test for `longest common subsequence`, has that made it into a package?  I could really use that right now.
